### PR TITLE
Add support for month & year intervals

### LINKED ISSUE #2796: Unable to work with month intervals
**Describe the bug**
When performing queries such as:

```
select date '1994-01-01' + interval '1' month as date;
```

I observe the following error:

```
`Err` value: "Execution(\"DateIntervalExpr does not support IntervalYearMonth\") at Optimizing logical plan for...
```

**To Reproduce**
Run the above query.

**Expected behavior**
`1994-02-01`

**Additional context**
Impact: Missing the month & year intervals are causing the TPC-H benchmarks to fail.

### PR DESCRIPTION
**Pull request by avantgardnerio**
# Which issue does this PR close?

Closes #2796.

 # Rationale for this change
In order to pass TPC-H benchmarks, datafusion will need to be able to support month & year date intervals.

# What changes are included in this PR?

A refactoring of the datetime module to:
1. Support `IntervalYearMonth` in addition to the currently supported `IntervalDayTime`
2. Perform some non-trivial date modulo math (that I hopefully got correct)
3. Add tests to ensure modulo math edge cases work correctly
4. Switch to an arguably more readable [railroad error handling pattern](https://fsharpforfunandprofit.com/rop/) to take better advantage of the error propagation (`?`) operator.

# Are there any user-facing changes?
More queries should work. Nothing that worked previously should break.

### CODE REVIEW

File: datafusion/physical-expr/src/expressions/datetime.rs [hunk 1]
```diff
@@ -86,76 +89,114 @@ impl PhysicalExpr for DateIntervalExpr {
...
+        // Add interval
+        let posterior = match scalar {
+            ScalarValue::IntervalDayTime(Some(_)) => {
+                prior.add(chrono::Duration::days(interval))
+            }
+            ScalarValue::IntervalYearMonth(Some(_)) => {
+                let target = add_months(prior, interval);
+                let target_plus = add_months(target, 1);
+                let last_day = target_plus.sub(chrono::Duration::days(1));
+                let day = min(prior.day(), last_day.day());
```
Comment (alamb) [inline datafusion/physical-expr/src/expressions/datetime.rs:157]: I don't really understand this logic -- I would have expected that code would split the interval into the 3 three fields (month, day, nano) and then use the various Chrono operations to do the timestamp arithmetic

https://github.com/apache/arrow/blob/master/format/Schema.fbs#L354-L375

File: datafusion/physical-expr/src/expressions/datetime.rs [hunk 2]
```diff
@@ -86,76 +89,114 @@ impl PhysicalExpr for DateIntervalExpr {
...
+            _ => {
+                // this should be unreachable because we check the operators in `try_new`
+                Err(DataFusionError::Execution(
+                    "Invalid operator for DateIntervalExpr".to_string(),
+                ))?
+            }
+        };
+
+        // Add interval
+        let posterior = match scalar {
```
Comment (alamb) [inline datafusion/physical-expr/src/expressions/datetime.rs:149]: It would b great to eventually put some/all of this logic into the arrow-rs kernels -- for example

https://docs.rs/arrow/16.0.0/arrow/compute/kernels/temporal/index.html

or perhaps https://docs.rs/arrow/16.0.0/arrow/compute/kernels/arithmetic/fn.add.html

That would also likely result in support for columnar execution support (aka adding a column of integers)

Maybe we can (I will file a ticket) start with kernels in datafusion and then port them to arrow.rs

**Review by alamb — COMMENTED** [review]
Thank you for the contribution @avantgardnerio -- this is looking like a great start. I don't fully understand some of the calculations but I may be misunderstanding

cc @ovr

File: datafusion/physical-expr/src/expressions/datetime.rs (hunk 1 above)
Comment (avantgardnerio) [reply datafusion/physical-expr/src/expressions/datetime.rs:157]: When the type is `IntervalDayTime`, I think I copied the behavior of the existing code, which was to grab a 32-bit number, which appears to represent the number of days based on this snippet https://github.com/apache/arrow-datafusion/blob/7617d78809d4ff5bde31142e0744c70024e40635/datafusion/physical-expr/src/expressions/datetime.rs#L122 (I assume there were tests around this? and if so they still pass).

For `IntervalYearMonth` I guessed and was correct based on the documentation you linked (ty!):
```
YEAR_MONTH - Indicates the number of elapsed whole months
```

I wasn't clear on what `IntervalMonthDayNano` meant, but now I do thanks to your link! This PR leaves it unimplemented (no change in behavior). It isn't critical to passing the TPC-H benchmarks, so it wasn't a priority for me at this moment.

As for the actual business logic of `IntervalYearMonth`, it is always stored as a number of months (year=12), and it turns out that adding (and subtracting) months is not very straight forward, and not implemented in chrono https://stackoverflow.com/questions/64081289/how-do-i-add-a-month-to-a-chrono-naivedate . When I looked into other implementations, they appeared to have issues (i.e. not supporting negative arguments), so I wrote this.

Unfortunately, one can't just add a fixed time unit (months have different numbers of days), and one has to know what is being added to in order to add properly (some Februaries have 29 days, others only 28). Observing postgres behavior, days of the month get clamped `2022-01-31` + 1 month = `2022-02-28`, so this logic is attempting to reproduce that behavior.

Comment (avantgardnerio) [reply datafusion/physical-expr/src/expressions/datetime.rs:157]: Reflecting, there could be a valid argument to PR chrono :thinking:

Comment (alamb) [reply datafusion/physical-expr/src/expressions/datetime.rs:157]: Thank you @avantgardnerio

Comment (avantgardnerio) [reply datafusion/physical-expr/src/expressions/datetime.rs]: I implemented `IntervalDayTime` and `IntervalMonthDayNano`, but could only figure out how to test `IntervalDayTime`, since I need a `timestamp` to test < 1 day resolutions, and it didn't seem like the timestamp path was plumbed or was plumbing it in scope for this PR.

File: datafusion/common/src/scalar.rs [hunk 3]
```diff
@@ -87,11 +89,14 @@ pub enum ScalarValue {
...
+    /// Number of elapsed whole months
     IntervalYearMonth(Option<i32>),
-    /// Interval with DayTime unit
+    /// Number of elapsed days and milliseconds (no leap seconds)
+    /// stored as 2 contiguous 32-bit signed integers
     IntervalDayTime(Option<i64>),
-    /// Interval with MonthDayNano unit
+    /// A triple of the number of elapsed months, days, and nanoseconds.
+    /// Months and days are encoded as 32-bit signed integers.
+    /// Nanoseconds is encoded as a 64-bit signed integer (no leap seconds).
```
Comment (avantgardnerio) [inline datafusion/common/src/scalar.rs:99]: The docs are unfortunately ambiguous as to what order these words are stored in.

File: datafusion/core/tests/sql/timestamp.rs [hunk 4]
```diff
@@ -814,3 +814,123 @@ async fn group_by_timestamp_millis() -> Result<()> {
...
+    assert_batches_eq!(expected, &results);
+
+    Ok(())
+}
+
+#[tokio::test]
+async fn add_interval_day() -> Result<()> {
+    let ctx = SessionContext::new();
+
+    let sql = "select date '1994-01-15' + interval '1' day as date;";
```
Comment (avantgardnerio) [inline datafusion/core/tests/sql/timestamp.rs:902]: `interval 1 day` seemed to cause it to use the `IntervalDayTime` type.

File: datafusion/optimizer/src/simplify_expressions.rs [hunk 5]
```diff
@@ -1963,10 +1963,10 @@ mod tests {
...
         // Note that constant folder runs and folds the entire
         // expression down to a single constant (true)
-        let expected = "Projection: Date32(\"18636\") AS CAST(totimestamp(Utf8(\"2020-09-08T12:05:00+00:00\")) AS Date32) + IntervalDayTime(\"123\")\
-            \n  TableScan: test";
+        let expected = r#"Projection: Date32("18636") AS CAST(totimestamp(Utf8("2020-09-08T12:05:00+00:00")) AS Date32) + IntervalDayTime("528280977408")
+  TableScan: test"#;
         let actual = get_optimized_plan_formatted(&plan, &time);
 
-        assert_eq!(expected, actual);
+        assert_eq!(actual, expected);
```
Comment (avantgardnerio) [inline datafusion/optimizer/src/simplify_expressions.rs:1970]: JetBrains products assume `actual` is on the left when visualizing diffs. The `assert_eq!` macro seems to not specify, so I updated it to look correct in my IDE so as to be less confusing.

File: datafusion/optimizer/src/simplify_expressions.rs [hunk 6]
```diff
@@ -1951,7 +1951,7 @@ mod tests {
         let date_plus_interval_expr = to_timestamp_expr(ts_string)
             .cast_to(&DataType::Date32, schema)
             .unwrap()
-            + Expr::Literal(ScalarValue::IntervalDayTime(Some(123)));
```
Comment (avantgardnerio) [inline datafusion/optimizer/src/simplify_expressions.rs:1954]: Based upon tests that used the parser to load "real" dates seemed to confirm the lower-order bytes are the `ms` part, so this was adding `123ms` which wasn't enough to bump the 32 bit date forward. Since the parser based tests are more likely to be correct than this hand-rolled one, I updated this to match that.

File: datafusion/physical-expr/src/expressions/datetime.rs [hunk 7]
```diff
@@ -74,88 +77,123 @@ impl PhysicalExpr for DateIntervalExpr {
...
+
+fn add_months(prior: NaiveDate, interval: i32) -> NaiveDate {
+    let target = chrono_add_months(prior, interval);
+    let target_plus = chrono_add_months(target, 1);
+    let last_day = target_plus.sub(chrono::Duration::days(1));
+    let day = min(prior.day(), last_day.day());
+    NaiveDate::from_ymd(target.year(), target.month(), day)
+}
+
+fn chrono_add_months(dt: NaiveDate, delta: i32) -> NaiveDate {
```
Comment (avantgardnerio) [inline datafusion/physical-expr/src/expressions/datetime.rs:191]: I still think I would like to PR this to chrono when I get a chance, and we can remove it from datafusion.

File: datafusion/physical-expr/src/expressions/datetime.rs [hunk 8]
```diff
@@ -74,88 +77,123 @@ impl PhysicalExpr for DateIntervalExpr {
...
+            _ => Err(DataFusionError::Execution(format!(
+                "Invalid lhs type for DateIntervalExpr: {}",
+                scalar
+            )))?,
+        };
+        Ok(res)
     }
 }
+
+fn add_m_d_nano(prior: NaiveDate, interval: i128, sign: i32) -> NaiveDate {
```
[the remark this answers is not in the document]
Comment (alamb) [reply datafusion/physical-expr/src/expressions/datetime.rs:163]: Perhaps we can add some unit tests in this module:

```rust
[#cfg(test)]
mod test {
  [#test] 
  fn test_calculations() { ... }
}
```

File: datafusion/physical-expr/src/expressions/datetime.rs (hunk 7 above)
Comment (alamb) [reply datafusion/physical-expr/src/expressions/datetime.rs:189]: 👍  - I like filing a ticket / PR in the target repo and then leaving a link in the comments of DataFusion

Btw I poked around and found this code in arrow2 (from @jorgecarleitao ) that is similar: https://docs.rs/arrow2/latest/src/arrow2/temporal_conversions.rs.html#342-368

**Review by alamb — COMMENTED** [review]
Thanks @avantgardnerio  I think this PR's code is looking good

The only thing I think it needs before merging are tests covering more of the date/time conversion logic. Once that is done I can write up a "move this to arrow" type ticket as well

**Comment by avantgardnerio** [comment]
Sorry this stalled out. I was really close on the subqueries, and I think I have them all working now, so I'll jump back on this.

**Comment by avantgardnerio** [comment]
Added unit tests. Will file PRs to respective dependency repos as well.

- chrono - I can figure out where to send the PR
- arrow - any idea what file / impl the functions listed as `TODO` should go into?

**Comment by avantgardnerio** [comment]
I [PRed](https://github.com/chronotope/chrono/pull/731) the `chrono` repo.

**Comment by avantgardnerio** [comment]
> Once that is done I can write up a "move this to arrow" type ticket 

@alamb , do you know where (in arrow), I can PR these functions to:

- `fn create_day_time(days: i32, millis: i32) -> i64`
- `fn create_month_day_nano(months: i32, days: i32, nanos: i64) -> i128`

?

**Comment by avantgardnerio** [comment]
> or perhaps https://docs.rs/arrow/16.0.0/arrow/compute/kernels/arithmetic/fn.add.html

Okay, I think I understand this. Ideally `math_op` would look something like this: ```
pub fn math_op<LT, RT, F>(
    left: &PrimitiveArray<LT>,
    right: &PrimitiveArray<RT>,
    op: F,
) -> Result<PrimitiveArray<LT>>
where
    LT: ArrowNumericType,
    RT: ArrowNumericType,
    F: Fn(LT::Native, RT::Native) -> LT::Native,
{
```
So that we could add an array of `IntervalXXX` to and array of `DateXXX`

**Comment by avantgardnerio** [comment]
What this is looking like when done inside `arrow-rs`: https://github.com/apache/arrow-rs/compare/master...spaceandtimelabs:arrow-rs:bg_date_math

**Comment by alamb** [comment]
> @alamb , do you know where (in arrow-rs), I can PR these functions to:


I recommend https://github.com/apache/arrow-rs/blob/master/arrow/src/compute/kernels/temporal.rs

**Comment by alamb** [comment]
For what it is worth https://github.com/apache/arrow-rs/compare/master...spaceandtimelabs:arrow-rs:bg_date_math looks pretty good to me 👍

File: datafusion/physical-expr/Cargo.toml [hunk 9]
```diff
@@ -44,6 +44,7 @@ arrow = { version = "17.0.0", features = ["prettyprint"] }
 blake2 = { version = "^0.10.2", optional = true }
 blake3 = { version = "1.0", optional = true }
 chrono = { version = "0.4", default-features = false }
+chronoutil = "0.2.3"
```
Comment (alamb) [inline datafusion/physical-expr/Cargo.toml:47]: I am somewhat concerned about adding dependencies (and this particular crate seems to be maintained by a single person) -- however,  since it is so small (and MIT licensced), we could also just inline the functions we care about.

Any thoughts @andygrove or @thinkharderdev ?

https://github.com/olliemath/chronoutil/blob/main/src/delta.rs

**Review by alamb — APPROVED** [review]
I think this is looking quite good -- thank you @avantgardnerio . I like the idea of getting the test coverage and then expanding support by adding kernels to arrow.

File: datafusion/physical-expr/Cargo.toml (hunk 9 above)
Comment (thinkharderdev) [reply datafusion/physical-expr/Cargo.toml:47]: I haven't reviewed the PR so am not sure how much of that create is used but I agree if it is a small amount of code it is better to just copy the code rather than add another dependency.

Comment (avantgardnerio) [reply datafusion/physical-expr/Cargo.toml]: I inlined it in my latest push. If we want to go back to including it in cargo.toml, it's an easy revert.

My $0.02 would be to leave it in cargo.toml, so that we can get updates if the author fixes something, and so that RAT checkers can see declaratively that we depend on that code, rather than trying to use heuristics of sematic code or binary diff.

**Review by alamb — APPROVED** [review]
I think this is good to go from my perspective. Does anyone else care to review?

cc @ovr  this may be interesting / relevant to your project 
cc @paddyhoran

**Review by andygrove — APPROVED** [review]

**Comment by alamb** [comment]
🐱 🚀
